### DISCUSSION TITLE
Web framework

### DISCUSSION THREAD

**Discussion by davidbrochart** [marimo-team/marimo#576] [Q&A] [answered]
I was wondering what were the reasons behind choosing Tornado for the server, as it's not a particularly modern technology compared to e.g. FastAPI.

**Comment by akshayka** [comment]
Thanks for the question @davidbrochart. The honest answer is that I saw that both Jupyter and streamlit used tornado, and it if it was good enough for them I figured it would be good enough for us.

Since we're still early, it's possible we could make the switch from tornado to FastAPI, or another framework, without too much effort. Do you think that would be worth it?

**Reply by davidbrochart** [reply]
The issue with Tornado is that it's not [ASGI](https://asgi.readthedocs.io)-compatible, and so it cannot access this ecosystem. Also, being a "legacy" technology, I don't think it's going to be maintained in the long run, so there's a risk there.
Jupyter is transitioning to FastAPI with [Jupyverse](https://github.com/jupyter-server/jupyverse). But [Litestar](https://github.com/litestar-org/litestar) is looking good too.

**Reply by akshayka** [reply]
Thanks for the links and context. Seems like it makes sense for us to switch to a more modern framework; we'll explore it.

**Reply by akshayka** [reply]
@davidbrochart, we [started a migration from tornado to FastAPI](https://github.com/marimo-team/marimo/pull/606) and it seems promising.

We noticed that the migration does bring in a few new packages, some of which are not yet stable, namely FastAPI and uvicorn. Pydantic is also brought in.

Like Jupyter, marimo is installed in users' virtual environments so we'll need to support a broad range of versions for these packages. 

Did you have any concerns on relying on these not-yet-stable packages while developing Jupyverse, and if so how did you mitigate them? Similarly, were there any concerns with supporting a wide range of Pydantic versions?

**Reply by davidbrochart** [reply]
Although FastAPI, Starlette or uvicorn didn't reach v1 yet, I think they are quite stable already. Also, since they are ASGI-compatible, you can swap uvicorn with [another ASGI server](https://github.com/florimondmanca/awesome-asgi?tab=readme-ov-file#servers), or FastAPI with [another application framework](https://github.com/florimondmanca/awesome-asgi?tab=readme-ov-file#application-frameworks) (this one will require changes in your code though). That said, FastAPI will [reach v1 soon](https://www.youtube.com/watch?v=QzuGFU6n_Gs&t=1376s).
I suggest to only support Pydantic v2, since v1 has a very different API (see [this discussion](https://github.com/pydantic/pydantic/discussions/5402)).

**Reply by akshayka** [reply]
Thanks for the links. The issue is that a lot of code still depends on Pydantic v1. I don't think we can make our users upgrade. For example, see what happened when langchain tried to force Pydantic v2 on their users: https://github.com/langchain-ai/langchain/discussions/9337

**Reply by davidbrochart** [reply]
I don't know how Marimo is designed, but if the user code runs in a different process/environment than the server, then this constraint wouldn't apply to user code.

**Comment by mscolnick** [comment, marked as answer]
We've migrated to use Starlette in [this PR](https://github.com/marimo-team/marimo/pull/606). Starlette is what FastAPI uses under the hood. (essentially the same without DI, decorators, and Pydantic build in).

marimo is now in the future serving via an ASGI framework.  

Thank you for the push @davidbrochart!
